### DISCUSSION TITLE
Include files are not linked at /usr/local/include or /usr/include

### DISCUSSION THREAD

**Discussion by watjurk** [Homebrew/discussions#868] [Everyday usage]
Mac: M1 Apple Silicon, BigSur: 11.2.1
Brew: Homebrew 3.0.1-116-g3ff5795

I don't know if this is intended or not but /usr/local/include and /usr/include does not exist on my system. Brew don't care and the only include path with brew installed formulas is at /opt/homebrew/include.

Is there any way to get brew to "link" includes at /usr/local/include or /usr/include ?

Thanks !

**Comment by Bo98** [comment]
This is intentional for Apple Silicon.

> Is there any way to get brew to "link" includes at /usr/local/include or /usr/include ?

`brew` does not support doing this. What issue are you having that makes you feel like you need this?

**Reply by Syzygies** [reply]
I'll describe an issue that brought me to this discussion. I do not need help with a particular issue; one can find workarounds each time. I would like to understand a "once and for all" solution.

I maintain a programming language zoo, and many of these builds take place outside Homebrew, but can take advantage of libraries provided most easily by Homebrew. I am at the moment setting up three Macs at once, in both flavors, and working around the M1 differences.

With some tweaking, one can base a native M1 build of Idris2 on Racket's native M1 build of Chez Scheme. The Idris2 build requires `gmp`. The build uses `/usr/bin/cc` which knows about `/usr/local/include` but not `/opt/homebrew/include`.  One needs to set the `CPATH` environment variable on an M1 Mac, but not on an Intel Mac.

I have a reasonable expectation that external builds should be able to find Homebrew-installed libraries on either M1 or Intel Macs, without per-build intervention. `eval $(${HOMEBREW_PREFIX}/bin/brew shellenv)` does not suffice. What am I missing?

(An aside, definitely not worth its own thread, but why does `brew shellenv` put a trailing colon on `MANPATH` but not `PATH` or `INFOPATH`? I spent twenty minutes thinking I was being clumsy; I find there's a very thin line between understanding my code and falling off the cliff, so I chase everything like this. Is there a reason for this lack of parallelism in the code, or should I provide a pull request to remove that character?)

**Reply by carlocab** [reply]
> I have a reasonable expectation that external builds should be able to find Homebrew-installed libraries on either M1 or Intel Macs, without per-build intervention. `eval $(${HOMEBREW_PREFIX}/bin/brew shellenv)` does not suffice. What am I missing?

I guess the difficulty is that build systems often defy reasonable expectations. Not because build system developers like being difficult (although there are many times where it seems like they do): the issue is that different users often have mutually exclusive expectations, all of which could be deemed reasonable depending on the context.

If you don't want to use processor-dependent build directives, you can always unconditionally add `$(brew --prefix)/include` to your `CPATH` (or the appropriate equivalent to `CFLAGS`/`CXXFLAGS`/`CPPFLAGS`) and `-L$(brew --prefix)/lib` to your `LDFLAGS` (or whatever is the equivalent of these for the build system you use). These will resolve to the correct values as long as you're using Homebrew-provided libraries.

> why does `brew shellenv` put a trailing colon on `MANPATH` but not `PATH` or `INFOPATH`?

Because the semantics of `MANPATH` are different than that of `PATH` (and perhaps `INFOPATH`, but I haven't checked this). A trailing `:` in `MANPATH` means "append the default `MANPATH` specified by the system configuration files. A trailing `:` in `PATH` means append the current working directory to `PATH`, which isn't always a good idea.

**Reply by andrews05** [reply]
I ran into this issue when trying to build an existing program that relied on libpng. In addition to importing `<png.h>`, the make file used `-lpng` to link to the library. Under Intel, I just did `brew install libpng` and then it all just worked when I ran `make`.
Under M1, I ended up adding the following lines to my profile in order to get it to work:
```
export CPATH="$HOMEBREW_PREFIX/include:$CPATH"
export LIBRARY_PATH="$HOMEBREW_PREFIX/lib:$LIBRARY_PATH"
```
Is there a reason the homebrew shellenv doesn't/shouldn't set these itself? I wouldn't expect it to set CFLAGS or similar, but these paths certainly seem reasonable.

**Reply by song6174** [reply]
Same question. Homebrew install libpng current is installed into `/opt/homebrew/Cellar/libpng/1.6.44`. I have to specify specific link for each library I installed through homebrew. Is there a more convenient way to do this?

**Reply by gromgit** [reply]
> I have to specify specific link for each library I installed through homebrew.

That's only true for _keg-only_ formulae, which aren't automatically linked into the main Homebrew directories. The caveats for those formulae tell you quite clearly [a] that they're keg-only, and [b] the paths you need to configure for use (and they NEVER include the version numbers you mentioned above).

But for the vast majority of library formulae (and this includes `libpng`), you simply need to add `/opt/homebrew/include` to your build include path, and `/opt/homebrew/lib` to your build library path. That's about as convenient as it gets, and to answer @andrews05's question about why Homebrew doesn't automagically configure the appropriate environment variables:

1. different build systems may use different variables for this purpose
2. _removing_ automatically-added directories from your build configuration (when you really don't want them there) is often MUCH LESS convenient than simply adding them when needed

**Comment by pinheadmz** [comment]
I am working with this issue as well after upgrading and migrating. I work on a few nodejs projects that use node-gyp to build native addons to nodejs. It's super inconvenient to add compiler flags to each of those dependencies' binding.gyp files. If I could add something like `export CFLAGS -L$(brew --prefix)/lib` to my `~/.bash_profile` script I would be ok with that, but I dont think that will work

**Reply by carlocab** [reply]
Why won't it work?
